lib/asciidoctor/substitutors.rb

```diff
@@ -303,28 +321,19 @@ def extract_inner_passthrough text, pre, attributes = nil
   # Internal: Restore the passthrough text by reinserting into the placeholder positions
   #
   # text  - The String text into which to restore the passthrough text
-  # outer - A Boolean indicating whether we are in the outer call (default: true)
   #
   # returns The String text with the passthrough text restored
-  def restore_passthroughs text, outer = true
+  def restore_passthroughs text
     passes = @passthroughs
-    # passthroughs may have been eagerly restored (e.g., footnotes)
-    #if outer && (passes.empty? || !text.include?(PASS_START))
-    #  return text
-    #end
-
     text.gsub PassSlotRx do
       # NOTE we can't remove entry from map because placeholder may have been duplicated by other substitutions
-      pass = passes[$1.to_i]
+      pass = passes[$1.to_i] || PASS_UNRESOLVED
```

Comment:
I don't really like this placeholder value. Is it to avoid code duplication ? 
I guess we could replace it with a condition ? if `passes` is defined then do X else to do Z ?

Reply:
This should never happen. But in the extreme case that it does (perhaps a misbehaving extension), we just want to avoid a crash. And because it's such an exception, I don't want to waste any time checking if the value is there. (We don't have to use a constant...we could just inline the value).

Reply:
Is your issue with the constant or the fallback? If it's just a constant, I can inline it.

I prefer the fallback right now because we need to stop being bitten by this crash and this solves it with the least amount of overhead.

Reply:
It's both. 
As you've mentioned it's unlikely that `passes[$1.to_i]` will equal `{ text: '??pass??' }` but if it happens we won't know for sure if it's the fallback value or not (but that's not my biggest concern)

> And because it's such an exception, I don't want to waste any time checking if the value is there.

Behind the scene Ruby will check if the value is there right ? When you say "I don't want to waste any time checking" you mean at runtime ?
I think it should maybe be a bit faster since we don't need to check `if (type = pass[:type])` if `passes[$1.to_i]` is nil.
Also since we know that `pass[:subs]` will be undefined maybe we can skip this call `subbed_text = apply_subs(pass[:text], pass[:subs])`.

My point is that we execute the same code path but it fells like we should not.

```rb
if (pass = passes[$1.to_i])
  # do X
else
  # do Z 
end
```

Reply:
Let me text perf. As I mentioned, this is tough because this code is just hammered to death.

Reply:
Done. And you'll like that we even log an error now ;)